### PR TITLE
(Real™ package™)™

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
Replaces: https://github.com/emberjs/ember-string/pull/345

BREAKING:
- requires ember-auto-import or embroider
- drop support for ember-source < 3.28
- drop support for ember-cli < 3.28

Maybe breaking:
- Drop support for TS 4.9
  - We haven't declared a [rolling support policy](https://www.semver-ts.org/formal-spec/5-compiler-considerations.html?highlight=rolling#rolling-support-windows)

Overall output:
```bash
❯ cat dist/index.js | wc -l
     142

❯ cat declarations/index.d.ts | wc -l
     160
```

A note on the diff:
image

image

--------

a type=module packages is more of a real package than a v2 addon (currently) because v2 addons are often CJS masquerading as ESM. When a package.json _does not_ have `type=module` specified, it's assumed that the `.js` extension means "Common JS" / require. This isn't great if we want to be better integrated with broader ecosystem tooling. So, since `@ember/string` does not depend on any ember APIs, it's actually a great candidate for using real ESM, ahead of auto-import/embroider releasing fixes for ESM "ember-addon" packages. (An ember-addon is a package with `ember-addon` in its keywords list)

_Where'd the test-app go?_

I deleted it. This work is under the assumption that, because we don't use anything _from_ ember, and because vitest knows nothing about ember, some basic unit tests ensure that the package works, and would be compatible with all versions of ember that have ember-auto-import or are using embroider.

### CODE REVIEW

File: .github/workflows/push-dist.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        with:
+          node-version: 18
+          cache: pnpm
+      - name: Install Dependencies
+        run: pnpm install --frozen-lockfile
+      - uses: kategengler/put-built-npm-package-contents-on-branch@v2.0.0
+        with:
+          branch: dist
+          token: ${{ secrets.GITHUB_TOKEN }}
+          working-directory: 'ember-string'
```
Comment (NullVoxPopuli) [inline .github/workflows/push-dist.yml:33]:
```suggestion
```

File: package.json [hunk 2]
```diff
@@ -2,126 +2,82 @@
...
+    "start:js": "rollup --config --watch --no-watch.clearScreen",
+    "start:types": "tsc --declaration --watch",
+    "test": "cd tests && pnpm test",
+    "test:types": "tsc --noEmit --project ./tests/type-tests/tsconfig.json",
+    "test:watch": "cd tests && pnpm test:watch",
+    "prepack": "rollup --config"
   },
   "dependencies": {
-    "ember-cli-babel": "^7.26.6"
+    "decorator-transforms": "^1.0.1"
```
Comment (NullVoxPopuli) [inline package.json:47]: TODO: remove this, as there are no decorators here

File: package.json [hunk 3]
```diff
@@ -2,126 +2,82 @@
...
-      "tokenRef": "GITHUB_AUTH"
-    },
-    "npm": {
-      "publish": false
-    }
-  },
   "volta": {
-    "node": "12.22.3",
-    "yarn": "1.22.11"
+    "node": "20.11.1",
```
Comment (mkszepp) [inline package.json:80]: Node is here v20, but in ci scripts v18... maybe its better to hold sync

File: babel.config.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
+  "plugins": [
+  ["@babel/plugin-transform-typescript", { "allExtensions": true, "onlyRemoveTypeImports": true, "allowDeclareFields": true }],
+    "@embroider/addon-dev/template-colocation-plugin",
+    ["babel-plugin-ember-template-compilation", {
+      "targetFormat": "hbs",
+      "transforms": []
+    }],
+    ["module:decorator-transforms", { "runtime": { "import": "decorator-transforms/runtime" } }],
```
Comment (mkszepp) [inline babel.config.json:9]: Should be removed, as decorator is not needed

File: package.json (hunk 2 above)
Comment (NullVoxPopuli) [reply package.json]: removed

File: babel.config.json (hunk 4 above)
Comment (NullVoxPopuli) [reply babel.config.json]: thanks! removed

File: package.json (hunk 3 above)
Comment (NullVoxPopuli) [reply package.json:77]: good point -- node is actually irrelevant now, so I can clean up CI a bit, let it be defaults.

**Pull request merged**